### PR TITLE
Fixed retaining legend state in paired result graph during reflow.

### LINKED ISSUE #107: Paired result legend doesnot retain state during reflow after moving to no data view
# Story Description

As a developer of terra-graphs, I want to figure out the root cause of the below bug report

So that I can either go ahead and fix the issue (if it's an issue that will take a day or two) or log a new issue with details on how to fix the issue

We'll know we're done when the issue is either resolved, or a new issue is logged with details on how to fix the issue.

# Bug Report
Paired result legend state does not retain the state when we move to a no data view and return to a view that has data.

## Description

if we toggle off a legend state and click on panning and switch to a no data view, and pan back to the view which has data, legend state is not retained.

### Graphs affected

* [ ]  Bar graph
* [ ]  Bubble graph
* [ ]  Gantt chart
* [ ]  Line graph
* [X]  Paired result
* [ ]  Pie chart
* [ ]  Scatter plot
* [ ]  Timeline

## Link to CodeSandbox
Use [this](https://codesandbox.io/s/ecstatic-heisenberg-5tugf?file=/src/graph.js) template to reproduce the behavior.

### PR DESCRIPTION
**Pull request by AshishMotanamGurunadham**
### Summary

- Legend state is not getting carried during reflow in paired result graph. Made changes to paired result reflow legend logic to fix this issue.

Closes #107 

### Deployment Link

https://terra-graphs-deployed-pr-#.herokuapp.com/

### Testing

- Wrote a test case that tests the state of legend during different stages of reflow.
### Additional Details

Thank you for contributing to Terra.
@cerner/terra
@cerner/carbon

### CODE REVIEW

File: packages/carbon-graphs/src/js/controls/PairedResult/helpers/helpers.js [hunk 1]
```diff
@@ -409,6 +409,23 @@ const processDataPoints = (graphConfig, dataTarget, reflow = false) => {
...
+          !utils.hasValue(
+            graphConfig.shownTargets,
+            `${dataTarget.key}_${t}`,
+          )
+        ) {
+          graphConfig.shownTargets.push(`${dataTarget.key}_${t}`);
+        }
+      }
+    });
+  }
```
Comment (yuderekyu) [inline packages/carbon-graphs/src/js/controls/PairedResult/helpers/helpers.js:428]: here's a high level suggestion to be more declarative and have a more functional mindset, which results in easier to read code with less side effects. I extracted the logic into its own helper that accepts the necessary inputs, converted the forEach + push into a filter call, and made use of the spread operator to combine the final array. The spread will need to be changed if graphConfig is declared with a `const`. let me know what you think.
```js
// util defined outside of processDataPoints. needs a better name
const geValues = (reflow, dataTarget, label) => {
  if (!reflow && !dataTarget.values.length < 1) {
      const types = ['high', 'mid', 'low'];
      const shownTargets = types.filter(type => {
        const label = getValue(dataTarget.label, type);
        const hasValue = utils.hasValue(graphConfig.shownTargets, `${dataTarget.key}_${type}`)
        if (label && label.display && !hasValue) {
            return `${dataTarget.key}_${type}`;
          }
        }
      });

    return shownTargets;
  }
  return [];
}
 
const shownTargets = getValues(reflow, dataTarget, label);
graphConfig.showTargets = [...graphConfig.showTargets, ...shownTargets];
```

Comment (AshishMotanamGurunadham) [reply packages/carbon-graphs/src/js/controls/PairedResult/helpers/helpers.js:428]: @yuderekyu Your idea sounds good, but [processDataPoint](https://github.com/cerner/terra-graphs/pull/111/files#diff-d04c4cf6f73329ea29bf2a081e4f7b5d5ddf52580b58d6ce58bf9c7b2728fb1eR385) method is used to initialize some variables before generating a graph content. Showntargets is one of the variables that need to be initialized in that method, so to keep everything organized I just added this content in the method where it needs to be executed. And also it is used only in one place so I felt it is not necessary to have separate util. Correct me if i am wrong.

Comment (yuderekyu) [reply packages/carbon-graphs/src/js/controls/PairedResult/helpers/helpers.js:428]: sure, i'm just afraid that processDataPoint is doing way too much - splitting the logic into smaller transforms makes it easier to reason about, and less coupled - i think the whole function could make do with a re-work, which is probably out of scope of this. I do think it's best to leave the code better than what we left it, and here, we're adding another layer of complexity for this function. In this case, splitting the logic into separate function is more about being modular, and less about reuse

Comment (sdadn) [reply packages/carbon-graphs/src/js/controls/PairedResult/helpers/helpers.js:428]: We do have a future project to do a code overhaul for a major update that also cleans up functions and streamlines the logic. So I would suggest staying within the scope of the bug-fix for now and make a note of these suggestions for the next project.

**Review by sdadn — APPROVED** [review]

File: packages/carbon-graphs/src/js/controls/PairedResult/helpers/helpers.js (hunk 1 above)
Comment (yuderekyu) [reply packages/carbon-graphs/src/js/controls/PairedResult/helpers/helpers.js:428]: I'd argue that making code lean at the outset saves time and effort in the long run. You've introduced new logic here, why not spend the time to make it encapsulated, coding around the complex legacy code that's here? I won't stop you from merging this, but just food for thought. More often than not, "later" becomes "never" 🤷🏼

Comment (ryanthemanuel) [reply packages/carbon-graphs/src/js/controls/PairedResult/helpers/helpers.js:428]: +1 to @yuderekyu  suggestion (assuming it doesn't blow up the scope too much. I like it when we can try and modularize where we can and I do think processDataPoint is a little busy as it is.

Comment (AshishMotanamGurunadham) [reply packages/carbon-graphs/src/js/controls/PairedResult/helpers/helpers.js]: Corrected [here](https://github.com/cerner/terra-graphs/pull/111/commits/a4eabe34b7595d302599c25f50a034acf6d5550b#diff-d04c4cf6f73329ea29bf2a081e4f7b5d5ddf52580b58d6ce58bf9c7b2728fb1eR383) as needed.

**Review by ryanthemanuel — APPROVED** [review]

**Review by yuderekyu — APPROVED** [review]

**Review by benbcai — APPROVED** [review]

**Pull request merged**